extension/goplugin/environment.go

```diff
@@ -548,122 +589,125 @@ func (thisEnvironment *Environment) resourceFromContext(sch Schema, context map[
-	return GlobRawTypes[name]
+func (env *Environment) RawType(name string) reflect.Type {
+	return env.rawTypes[name]
 }
 
 // RegisterType registers a runtime type of resource for a given name
-func (thisEnvironment *Environment) RegisterType(name string, typeValue interface{}) {
+func (env *Environment) RegisterType(name string, typeValue interface{}) {
 	targetType := reflect.TypeOf(typeValue)
-	GlobTypes[name] = targetType
+	env.types[name] = targetType
 }
 
 // ResourceType returns a runtime type for a given named resource
-func (thisEnvironment *Environment) ResourceType(name string) reflect.Type {
-	return GlobTypes[name]
+func (env *Environment) ResourceType(name string) reflect.Type {
+	return env.types[name]
 }
 
 // Stop stops the environment to its initial state
-func (thisEnvironment *Environment) Stop() {
+func (env *Environment) Stop() {
 	log.Info("Stop environment")
 
-	// reset globals
-	GlobHandlers = nil
-	GlobSchemaHandlers = nil
-	GlobRawTypes = make(map[string]reflect.Type)
-	GlobEnvironments = map[string]*Environment{}
-
 	// reset locals
-	thisEnvironment.extCore = nil
-	thisEnvironment.extLogger = nil
-	thisEnvironment.extSchemas = nil
-	thisEnvironment.extSync = nil
-	thisEnvironment.extDatabase = nil
+	env.extCore = nil
+	env.extLogger = nil
+	env.extSchemas = nil
+	env.extSync = nil
+	env.extDatabase = nil
+
+	env.traceID = ""
+
+	env.handlers = nil
+	env.schemaHandlers = nil
+
+	env.rawTypes = make(map[string]reflect.Type)
+	env.types = make(map[string]reflect.Type)
+
+	// after stop
+	if env.afterStopHook != nil {
+		log.Debug("Calling after stop hook for go environment: %s", env.name)
+		env.afterStopHook(env)
+	} else {
+		log.Debug("After stop hook is not set for go environment: %s", env.name)
+	}
 }
 
 // Reset clear the environment to its initial state
-func (thisEnvironment *Environment) Reset() {
+func (env *Environment) Reset() {
 	log.Info("Reset environment")
 
-	thisEnvironment.Stop()
-	thisEnvironment.Start()
+	env.Stop()
+	env.Start()
 }
 
 // Clone makes a clone of the rawEnvironment
-func (thisEnvironment *Environment) Clone() extension.Environment {
-	env := &Environment{
-		source:          thisEnvironment.source,
-		beforeStartInit: thisEnvironment.beforeStartInit,
-
-		// internals
-		name:  thisEnvironment.name,
-		db:    thisEnvironment.db,
-		ident: thisEnvironment.ident,
-		sync:  thisEnvironment.sync,
-
-		// plugin related
-		manager: thisEnvironment.manager,
-		plugin:  thisEnvironment.plugin,
+func (env *Environment) Clone() extension.Environment {
+	clone := &Environment{
+		initFns:         env.initFns,
+		beforeStartHook: env.beforeStartHook,
+		afterStopHook:   env.afterStopHook,
 
-		schemasFnRaw: thisEnvironment.schemasFnRaw,
-		schemasFn:    thisEnvironment.schemasFn,
-		schemas:      thisEnvironment.schemas,
-
-		initFnRaw: thisEnvironment.initFnRaw,
-		initFns:   thisEnvironment.initFns,
+		name:  env.name,
+		db:    env.db,
+		ident: env.ident,
+		sync:  env.sync,
```

Comment:
You must clone handlers and schemaHandlers.

Reply:
right, fixed